### PR TITLE
Add docs for KEP 2200 (DenyServiceExternalIPs)

### PR DESCRIPTION
**Pull request by thockin**
* Document DenyServiceExternalIPs admission controller
* Re-order other admission controller blocks to be alphabetical
* Document DefaultIngressClass (missing)

KEP: https://github.com/kubernetes/enhancements/issues/2200
Code: https://github.com/kubernetes/kubernetes/pull/97395

### CODE REVIEW

**Comment by thockin** [comment]
Ugh, master not synced, will rebase.

**Comment by bells17** [comment]
/remove-language es fr id it ja ko zh

File: content/en/docs/reference/access-authn-authz/admission-controllers.md [hunk 1]
```diff
@@ -145,6 +145,22 @@ This admission controller observes creation of CertificateSigningRequest resourc
...
+request any special ingress class do not need to care about them at all and they will get the
+default one.
+
+This admission controller does not do anything when no default ingress class is configured. When more than one ingress
+class is marked as default, it rejects any creation of `Ingress` with an error and an administrator
+must revisit their `IngressClass` objects and mark only one as default (with the annotation
+"ingressclass.kubernetes.io/is-default-class").  This admission controller ignores any `Ingress`
+updates; it acts only on creation.
+
+See [ingress](/docs/concepts/services-networking/ingress/) documentation for more about ingress
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/admission-controllers.md:161]: nit:
```suggestion
See the [Ingress](/docs/concepts/services-networking/ingress/) documentation for more about ingress
```

File: content/en/docs/reference/access-authn-authz/admission-controllers.md [hunk 2]
```diff
@@ -169,33 +185,45 @@ have toleration for taints `node.kubernetes.io/not-ready:NoExecute` or
...
 {{< feature-state for_k8s_version="v1.13" state="deprecated" >}}
 
-This admission controller will deny exec and attach commands to pods that run with escalated privileges that
-allow host access.  This includes pods that run as privileged, have access to the host IPC namespace, and
-have access to the host PID namespace.
+This admission controller will intercept all requests to exec a command in a pod if that pod has a privileged container.
 
-The DenyEscalatingExec admission plugin is deprecated and will be removed in v1.18.
+This functionality has been merged into [DenyEscalatingExec](#denyescalatingexec).
+The DenyExecOnPrivileged admission plugin is deprecated and will be removed in v1.18.
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/admission-controllers.md:209]: Is this correct?

File: content/en/docs/reference/access-authn-authz/admission-controllers.md [hunk 3]
```diff
@@ -169,33 +185,45 @@ have toleration for taints `node.kubernetes.io/not-ready:NoExecute` or
...
 
 Use of a policy-based admission plugin (like [PodSecurityPolicy](#podsecuritypolicy) or a custom admission plugin)
 which can be targeted at specific users or Namespaces and also protects against creation of overly privileged Pods
 is recommended instead.
 
+### DenyServiceExternalIPs
+
+This admission controller rejects all net-new usage of the `Service` field `externalIPs`.  This
+feature is very powerful (allows network traffic interception) and not well policied.  When enabled,
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/admission-controllers.md:218]: Did you mean:
```suggestion
feature is very powerful (allows network traffic interception) and not well policed.  When enabled,
```
?

File: content/en/docs/reference/access-authn-authz/admission-controllers.md [hunk 4]
```diff
@@ -553,6 +581,37 @@ This admission controller also protects the access to `metadata.ownerReferences[
 of an object, so that only users with "update" permission to the `finalizers`
 subresource of the referenced *owner* can change it.
 
+### PersistentVolumeClaimResize {#persistentvolumeclaimresize}
+
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/admission-controllers.md:585]: nit: I know this is mostly a move, but I'd expect to see something like:
```markdown
{{< feature-state for_k8s_version="v1.XX" state="alpha" >}}
```
here.

File: content/en/docs/reference/access-authn-authz/admission-controllers.md [hunk 5]
```diff
@@ -553,6 +581,37 @@ This admission controller also protects the access to `metadata.ownerReferences[
...
+For example: all `PersistentVolumeClaim`s created from the following `StorageClass` support volume expansion:
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: gluster-vol-default
+provisioner: kubernetes.io/glusterfs
+parameters:
+  resturl: "http://192.168.10.100:8080"
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/admission-controllers.md:606]: nit: consider using a documentation IP address
```suggestion
  resturl: "http://192.0.2.100:8080"
```

Minor thing as 192.168.0.0/16 is private, and I can't see this misleading many people into using 192.168.10.100

**Review by sftim — COMMENTED** [review]
A few small items of feedback.

File: content/en/docs/reference/access-authn-authz/admission-controllers.md (hunk 2 above)
Comment (thockin) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md:209]: It's apparently still in the tree - apiserver still lists it.  Code still in k/k with @liggitt fingerprints on the deprecation.

File: content/en/docs/reference/access-authn-authz/admission-controllers.md (hunk 3 above)
Comment (thockin) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md:218]: I meant policied, but maybe that's an awkward word?  Will change it to "not well controlled by policy"

File: content/en/docs/reference/access-authn-authz/admission-controllers.md (hunk 5 above)
Comment (thockin) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md:606]: In real usage this will be a private IP of their gluster control plane.

**Comment by thockin** [comment]
nits fixed

File: content/en/docs/reference/access-authn-authz/admission-controllers.md (hunk 2 above)
Comment (liggitt) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md:209]: cleanup in https://github.com/kubernetes/kubernetes/pull/98669

**Comment by sftim** [comment]
/retitle Add docs for KEP 2200 (DenyServiceExternalIPs)

**Comment by sftim** [comment]
This looks ready for tech review (which, I'm hoping, is straightforward).

**Comment by thockin** [comment]
This will be a short review, I hope :)

@bowei for Ingress paragraph
@tallclair for externalIPs paragraph

/assign bowei
/assign tallclair

File: content/en/docs/reference/command-line-tools-reference/kube-apiserver.md [hunk 6]
```diff
@@ -470,7 +470,7 @@ kube-apiserver [flags]
...
...viceAccount, TaintNodesByCondition, Priority, DefaultTolerationSeconds, DefaultStorageClass, StorageObjectInUseProtection, PersistentVolumeClaimResize, RuntimeClass, CertificateApproval, CertificateSigning, CertificateSubjectRestriction, DefaultIngressClass, MutatingAdmissionWebhook, ValidatingAdmissionWebhook, ResourceQuota). Comma-delimited list of admission plugins: AlwaysAdmit, AlwaysDeny, AlwaysPullImages, CertificateApproval, CertificateSigning, CertificateSubjectRestriction, DefaultIngressClass, DefaultStorageClass, DefaultTolerationSeconds, DenyEscalatingExec, DenyExecOnPrivileged, DenyServiceExternalIPs, EventRateLimit, ExtendedResourceToleration, ImagePolicyWebhook, LimitPodHardAntiAffinityTopology, LimitRanger, MutatingAdmissionWebhook, NamespaceAutoProvision, NamespaceExists, NamespaceLifecycle, NodeRestriction, OwnerReferencesPermissionEnforcement, PersistentVolumeClaimResize, PersistentVolumeLabel, PodNodeSelector, PodSecurityPolicy, PodTolerationRestriction, Priority, ResourceQuota, RuntimeClass, SecurityContextDeny, ServiceAccount, StorageObjectInUseProtection, TaintNodesByCondition, ValidatingAdmissionWebhook. The order of plugins in this flag does not matter.</td>
```
Comment (kbhawkey) [inline content/en/docs/reference/command-line-tools-reference/kube-apiserver.md:473]: 👋 @thockin . This file is auto-generated at release time.
If the kube-apiserver `enable-admission-plugins` flag has been updated, this page will be regenerated with the correct list of plugins.

File: content/en/docs/reference/access-authn-authz/admission-controllers.md [hunk 7]
```diff
@@ -105,6 +105,12 @@ NamespaceLifecycle, LimitRanger, ServiceAccount, TaintNodesByCondition, Priority
 
 This admission controller allows all pods into the cluster. It is deprecated because its behavior is the same as if there were no admission controller at all.
 
+### AlwaysDeny {#alwaysdeny}
+
+{{< feature-state for_k8s_version="v1.13" state="deprecated" >}}
+
+Rejects all requests. AlwaysDeny is DEPRECATED as no real meaning.
```
Comment (kbhawkey) [inline content/en/docs/reference/access-authn-authz/admission-controllers.md:112]: nit: AlwaysDeny is DEPRECATED as it has no real meaning?

File: content/en/docs/reference/access-authn-authz/admission-controllers.md [hunk 8]
```diff
@@ -169,33 +185,46 @@ have toleration for taints `node.kubernetes.io/not-ready:NoExecute` or
...
 {{< feature-state for_k8s_version="v1.13" state="deprecated" >}}
 
-This admission controller will intercept all requests to exec a command in a pod if that pod has a privileged container.
+This admission controller will deny exec and attach commands to pods that run with escalated privileges that
+allow host access.  This includes pods that run as privileged, have access to the host IPC namespace, and
+have access to the host PID namespace.
 
-This functionality has been merged into [DenyEscalatingExec](#denyescalatingexec).
-The DenyExecOnPrivileged admission plugin is deprecated and will be removed in v1.18.
+The DenyEscalatingExec admission plugin is deprecated and will be removed in v1.18.
```
Comment (kbhawkey) [inline content/en/docs/reference/access-authn-authz/admission-controllers.md:196]: nit: The DenyEscalatingExec admission plugin is deprecated and has been removed in Kubernetes v1.18?

Comment (kbhawkey) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md:196]: Just saw the question below. It is not clear if this plugin will be deprecated in 1.21 or added to the v1.18 version as a patch?

File: content/en/docs/reference/command-line-tools-reference/kube-apiserver.md (hunk 6 above)
Comment (thockin) [reply content/en/docs/reference/command-line-tools-reference/kube-apiserver.md:473]: Should I remove it from this PR?

File: content/en/docs/reference/access-authn-authz/admission-controllers.md (hunk 8 above)
Comment (thockin) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md:196]: Yeah, I just moved this block around - if it gets removed, that is a different PR :)

File: content/en/docs/reference/command-line-tools-reference/kube-apiserver.md (hunk 6 above)
Comment (kbhawkey) [reply content/en/docs/reference/command-line-tools-reference/kube-apiserver.md:473]: It's not a problem to leave it. A new version of the file will be generated but may not be needed if the updates are the same.

File: content/en/docs/reference/access-authn-authz/admission-controllers.md [hunk 9]
```diff
@@ -94,7 +94,7 @@ kube-apiserver -h | grep enable-admission-plugins
 In the current version, the default ones are:
 
 ```shell
-NamespaceLifecycle, LimitRanger, ServiceAccount, TaintNodesByCondition, Priority, DefaultTolerationSeconds, DefaultStorageClass, StorageObjectInUseProtection, PersistentVolumeClaimResize, RuntimeClass, CertificateApproval, CertificateSigning, CertificateSubjectRestriction, DefaultIngressClass, MutatingAdmissionWebhook, ValidatingAdmissionWebhook, ResourceQuota
+AlwaysAdmit, AlwaysDeny, AlwaysPullImages, CertificateApproval, CertificateSigning, CertificateSubjectRestriction, DefaultIngressClass, DefaultStorageClass, DefaultTolerationSeconds, DenyEscalatingExec, DenyExecOnPrivileged, DenyServiceExternalIPs, EventRateLimit, ExtendedResourceToleration, ImagePolicyWebhook, LimitPodHardAntiAffinityTopology, LimitRanger, MutatingAdmissionWebhook, NamespaceAutoProvision, NamespaceExists, NamespaceLifecycle, NodeRestriction, OwnerReferencesPermissionEnforcement, PersistentVolumeClaimResize, PersistentVolumeLabel, PodNodeSelector, PodSecurityPolicy, PodTolerationRestriction, Priority, ResourceQuota, RuntimeClass, SecurityContextDeny, ServiceAccount, StorageObjectInUseProtection, TaintNodesByCondition, ValidatingAdmissionWebhook
```
Comment (tallclair) [inline content/en/docs/reference/access-authn-authz/admission-controllers.md:97]: This doesn't look right to me. Isn't this just all of them? I think the default list is defined here: https://github.com/kubernetes/kubernetes/blob/1119a505aca14467accedf850daf30aa9c532ef2/pkg/kubeapiserver/options/plugins.go#L143-L161

File: content/en/docs/reference/access-authn-authz/admission-controllers.md (hunk 2 above)
Comment (tallclair) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md:209]: Thanks Jordan. Let's update this to say v1.21 (same above)

File: content/en/docs/reference/access-authn-authz/admission-controllers.md (hunk 9 above)
Comment (thockin) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md:97]: Good catch.  I cut-and-pasted the wrong list.  That said, is it really worthwhile to list "in the current version" in docs - that doesn't help much.  What version?  Maybe we should just nix this block?

Comment (thockin) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md:97]: I'll update the list, but still think we should remove it totally.

Comment (tallclair) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md]: Yeah, I think just giving the command should be sufficient, or maybe even linking to the code. I could go either way.

Comment (tengqm) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md]: If I'm remembering this correctly, the order of this list matters. Is that still the case? I just noticed that the list is now in alphabetic order.

Comment (liggitt) [reply content/en/docs/reference/access-authn-authz/admission-controllers.md]: When using --enable-admission-plugins/--disable-admission-plugins, the order is unimportant. That just toggles them on/off, but the order is predetermined.

File: content/en/docs/reference/access-authn-authz/admission-controllers.md [hunk 10]
```diff
@@ -145,6 +145,22 @@ This admission controller observes creation of CertificateSigningRequest resourc
...
+request any special ingress class do not need to care about them at all and they will get the
+default one.
+
+This admission controller does not do anything when no default ingress class is configured. When more than one ingress
+class is marked as default, it rejects any creation of `Ingress` with an error and an administrator
+must revisit their `IngressClass` objects and mark only one as default (with the annotation
+"ingressclass.kubernetes.io/is-default-class").  This admission controller ignores any `Ingress`
+updates; it acts only on creation.
+
+See the [ingress](/docs/concepts/services-networking/ingress/) documentation for more about ingress
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/admission-controllers.md:161]: mini nit:
```suggestion
See the [Ingress](/docs/concepts/services-networking/ingress/) documentation for more about ingress
```

**Review by sftim — COMMENTED** [review]
/approve

**Pull request merged**